### DISCUSSION TITLE
Cancellable io_bound / cpu_bound calls

### DISCUSSION THREAD

**Discussion by frankvp11** [zauberzeug/nicegui#3347] [Ideas / Feature Requests]
This question is more so specific to the io_bound calls. 
I was wondering if there was a way to cancel them? 
Suppose you take the documentation example and based on some condition you wanted to cancel the io_bound call. How would you go about doing that - a good way to test this would be with a button test. Place another button which would cancel it.
As far as I'm aware there is no way to actually do this. 
However, 'stealing' the source code, you can do something like this:

```python

from nicegui import ui, run
import requests
import asyncio
from concurrent.futures import ThreadPoolExecutor
import requests
from functools import partial

class Page:
    def __init__(self) -> None:
        self.create()
        self.result = "none"
    
    def cancel_future(self):
        self.future.cancel()
        self.result = "Cancelled"

    def create(self):
        ui.button("Long call", on_click=self.long_call)
        ui.button("Cancel", on_click=self.cancel_future)
        ui.label().bind_text(self, "result")

    async def long_call(self):
        loop = asyncio.get_running_loop()
        self.future = loop.run_in_executor(
            ThreadPoolExecutor(),
            partial(requests.get, 'https://httpbin.org/delay/3', timeout=10)
        ) 
        response = await self.future
        self.result = len(response.content)

@ui.page("/")
def main():
    page = Page()
    

ui.run()
```

Is there some way to replicate this in the current NiceGUI implementation, and if not, would you (NiceGUI) be open to helping me find an implementation which allows you to cancel the futures?

**Comment by burnpanck** [comment]
In the particular case of `requests.get`, it would be much better to use a native async library, like [httpx](https://www.python-httpx.org/). In more general cases, you should look for "structured concurrency" (see e.g. the [background section of PEP0789](https://peps.python.org/pep-0789/#background)): a paradigm that guarantees cancelability.

Otherwise, with generic threads, there is basically no way to make cancellation work: In general, cancellation is hard: It's impossible to make it work generally without risking to leak resources, because cleaning-up requires active participation of the software that is being cancelled. Unfortunately, old-school asyncio doesn't help here much. While they recently introduced primitives which follow "structured concurrency" design pattern (where every task is responsible for cleaning up the resources it creates), it fails as soon as a single component does not follow it. Whenever you try to access the event loop using `get_running_loop` or similar, you know you're out of that pattern.
